### DISCUSSION TITLE
Add list-alsa-devices to identify how pass into device params when card has multiple devices

### DISCUSSION THREAD

**Discussion by cdgraff** [savonet/liquidsoap#2928] [User Support]
Please, can add the same work doit for pulse-audio into ALSA?

Call `liquidsoap --list-alsa-devices` to get the list of recognized devices
Use `input.alsa(device=xxx, ...)` where device is the ID of the device that you wish to use.

Same as #2733 but for ALSA DEVICES

**Comment by smimram** [comment]
You can use
```
aplay -l
```
for this. Would there be any advantage to having this builtin in Liquidsoap?

**Comment by toots** [comment]
Agree with @smimram here. We added the pulse device listing because there was no obvious way to do it.

I've said but I'll say it again, I personally consider ALSA to be a _really_ bad audio API.

It is kernel level with direct access to hardware. This is the list that `aplay -l` returns. That access is _super_ low-level. Like, no sample rate conversion, etc. You have to query the hardware only with what it can do. Also, access on hardware-level devices is exclusive so if one app is using a hardware device, it is locked for other applications.

Then, alsa also defines higher-level devices that _try_ to do good things like samplerate conversion, input or output sharing etc. This is the list that `aplay -L` returns. However, the nomenclature and automatic configuration of these devices is terrible and really hard to understand. It's divided in finer-grained named device that make it super hards to know which device to use etc.

Frankly, my approach with alsa is, either you know what you are doing and should directly query the right hardware/abstract device and probably won't need help from us or, else, please, use a higher-level API such as pulseaudio and etc. We try our best to implement our ALSA operators following the ALSA API but that API is so bad, it just pushes the issues up to our users and for that, we're really sorry about it.

**Reply by cdgraff** [reply]
ok I understand! My idea was after try to use

arecord -l
```
**** List of CAPTURE Hardware Devices ****
card 0: ASI88210 [ASI8821-0], device 0: Asihpi PCM [Asihpi PCM]
  Subdevices: 4/4
  Subdevice #0: subdevice #0
  Subdevice #1: subdevice #1
  Subdevice #2: subdevice #2
  Subdevice #3: subdevice #3
```

And got error passing `hw:0,0,0`

`liquidsoap 'output.icecast(%mp3, host = "xxxx", port = 9000, password = "xxx", mount = "/alsa-test", input.alsa(device="hw:0,0,0"))'`

```
2023/02/23 20:58:48 [input.alsa_0:3] Overrun!
2023/02/23 20:58:48 [input.alsa_0:2] Trying to recover..
```

Using default works well, but I need use the others too.

Some advice?

**Reply by cdgraff** [reply]
For future reference... and documentation about integration of Liquidsoap + Audioscience Tunner cards

Get the String from:
```
$ arecord -L
...
sysdefault:CARD=ASI88210
    ASI8821-0, Asihpi PCM
    Default Audio Device
dmix:CARD=ASI88210,DEV=0
    ASI8821-0, Asihpi PCM
    Direct sample mixing device
dsnoop:CARD=ASI88210,DEV=0
    ASI8821-0, Asihpi PCM
    Direct sample snooping device
hw:CARD=ASI88210,DEV=0
    ASI8821-0, Asihpi PCM
    Direct hardware device without any conversions
plughw:CARD=ASI88210,DEV=0
    ASI8821-0, Asihpi PCM
    Hardware device with all software conversions
usbstream:CARD=ASI88210
    ASI8821-0
    USB Stream Output
```


Need pass this way the reference:
**Tunner1**: `input.alsa(device="hw:CARD=ASI88210,DEV=0,0"))`
**Tunner2**: `input.alsa(device="hw:CARD=ASI88210,DEV=0,1"))`
**Tunner3**: `input.alsa(device="hw:CARD=ASI88210,DEV=0,2"))`
**Tunner4**: `input.alsa(device="hw:CARD=ASI88210,DEV=0,3"))`

**Comment by smimram** [comment]
That being said, for future ref, this would not be very difficult to add. The necessary functions are now there: https://github.com/savonet/ocaml-alsa/pull/3
